sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java

```diff
@@ -0,0 +1,101 @@
+ */
+public class AzureKeyVaultKeyWrapProvider extends EncryptionKeyWrapProvider {
+    private AzureKeyVaultKeyStoreProvider azureKeyVaultKeyStoreProvider;
+    private final static ImplementationBridgeHelpers.CosmosExceptionHelper.CosmosExceptionAccessor cosmosExceptionAccessor =
+        ImplementationBridgeHelpers.CosmosExceptionHelper.getCosmosExceptionAccessor();
+
+    public AzureKeyVaultKeyWrapProvider(TokenCredential tokenCredential) {
+        try {
+            this.azureKeyVaultKeyStoreProvider = new AzureKeyVaultKeyStoreProvider(tokenCredential);
+        } catch (MicrosoftDataEncryptionException ex) {
+            throw cosmosExceptionAccessor.createCosmosException(HttpConstants.StatusCodes.INTERNAL_SERVER_ERROR, ex);
+        }
+    }
+
+    /**
+     * Getter for name of the key vault provider.
+     *
+     * @return provider name
+     */
+    @Override
+    public String getProviderName() {
+        return this.azureKeyVaultKeyStoreProvider.getProviderName();
+    }
+
+    /**
+     * Decrypts an encrypted DEK with the RSA encryption algorithm using the asymmetric key specified by the key path
+     *
+     * @param encryptionKeyId                 - Complete path of an asymmetric key in Azure Key Vault
+     * @param cosmosKeyEncryptionKeyAlgorithm - Asymmetric Key Encryption Algorithm
+     * @param encryptedKey                    - Encrypted Data Encryption Key
+     * @return Plain text data encryption key
+     */
+    @Override
+    public byte[] unwrapKeyAsync(String encryptionKeyId, String cosmosKeyEncryptionKeyAlgorithm, byte[] encryptedKey) {
+        try {
+            if (!KeyEncryptionKeyAlgorithm.RSA_OAEP.equals(cosmosKeyEncryptionKeyAlgorithm)) {
+                throw new IllegalArgumentException("The specified KeyEncryptionAlgorithm is not supported. Please " +
+                    "refer to https://aka.ms/CosmosClientEncryption for more details. ");
+            }
+            return this.azureKeyVaultKeyStoreProvider.unwrapKey(encryptionKeyId, com.microsoft.data.encryption.cryptography.KeyEncryptionKeyAlgorithm.RSA_OAEP,
+                encryptedKey);
+        } catch (MicrosoftDataEncryptionException ex) {
+            throw cosmosExceptionAccessor.createCosmosException(HttpConstants.StatusCodes.INTERNAL_SERVER_ERROR, ex);
+        }
+    }
+
+    /**
+     * Encrypts a DEK with the RSA encryption algorithm using the asymmetric key specified by the key path.
+     *
+     * @param encryptionKeyId                 - Complete path of an asymmetric key in Azure Key Vault
+     * @param cosmosKeyEncryptionKeyAlgorithm - Asymmetric Key Encryption Algorithm
+     * @param key                             - Plain text data encryption key
+     * @return Encrypted data encryption key
+     */
+    @Override
+    public byte[] wrapKeyAsync(String encryptionKeyId, String cosmosKeyEncryptionKeyAlgorithm, byte[] key) {
```

Comment:
another question is why the method is async suffix? seems like there is no async operation in the method.

Reply:
good catch , removed async